bloop/models.py

```diff
@@ -191,23 +202,54 @@ def setup_columns(meta):
     }
 
 
+def clone_index(index):
+    if index.projection['mode'] == 'include':
+        projection = index.projection['included']
+    else:
+        projection = index.projection['mode']
+
+    if isinstance(index, GlobalSecondaryIndex):
+        return GlobalSecondaryIndex(
+            projection=projection, hash_key=index.hash_key, range_key=index.range_key,
+            read_units=index.read_units, write_units=index.write_units, dynamo_name=index._dynamo_name)
+    else:
+        return LocalSecondaryIndex(
+            projection=projection, range_key=index.range_key, dynamo_name=index._dynamo_name,
+            strict=index.projection['strict'])
+
+
 def setup_indexes(meta):
     """Filter indexes from fields, compute projection for each index"""
     # Don't put these in the metadata until they bind successfully.
     gsis = set()
     lsis = set()
-    for attr in meta.model.__dict__.values():
-        if isinstance(attr, GlobalSecondaryIndex):
-            gsis.add(attr)
-        if isinstance(attr, LocalSecondaryIndex):
-            lsis.add(attr)
+
+    for name, index in inspect.getmembers(meta.model, lambda x: isinstance(x, GlobalSecondaryIndex)):
+        if index.model is not meta.model:
+            index = clone_index(index)
+            index._name = name
+            setattr(meta.model, name, index)
+        gsis.add(index)
+    for name, index in inspect.getmembers(meta.model, lambda x: isinstance(x, LocalSecondaryIndex)):
+        if index.model is not meta.model:
+            index = clone_index(index)
```

Comment:
This looks like we're duplicating the work from `Index.__set_name__` because it isn't called; there's a similar line in `bind_index` which checks `if index.model is None`; that will only be true when `__set_name__` wasn't called, right?

Would it be easier to drop the `__set_name__` logic and move it all to `bind_index`?  Then there's one place to hook all the machinery up, and it provides an easy way for someone extending bloop to get the same behavior.  The docs then say "if you want to attach an index to a model, use `bloop.models.bind_index(model, index)` to properly configure the class and its subclasses".